src/plugins/intel_npu/src/backend/src/zero_remote_tensor.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "zero_remote_tensor.hpp"
+
+#include <ze_api.h>
+
+#include "intel_npu/al/config/common.hpp"
+#include "openvino/core/type/element_iterator.hpp"
+#include "openvino/runtime/intel_npu/remote_properties.hpp"
+#include "zero_utils.hpp"
+
+namespace {
+
+constexpr std::size_t STANDARD_PAGE_SIZE = 4096;
+
+}  // namespace
+
+namespace intel_npu {
+
+ZeroRemoteTensor::ZeroRemoteTensor(std::shared_ptr<ov::IRemoteContext> context,
+                                   std::shared_ptr<ZeroInitStructsHolder> init_structs,
+                                   const ov::element::Type& element_type,
+                                   const ov::Shape& shape,
+                                   const Config& config,
+                                   RemoteTensorType tensor_type,
+                                   RemoteMemoryType mem_type,
+                                   void* mem)
+    : RemoteTensor(context, element_type, shape),
+      _config(config),
+      _logger("ZeroRemoteContext", _config.get<LOG_LEVEL>()),
+      _init_structs(init_structs),
+      _tensor_type(tensor_type),
+      _mem_type(mem_type),
+      _mem(mem) {
+    _ze_properties.stype = ZE_STRUCTURE_TYPE_DEVICE_PROPERTIES;
+    zeroUtils::throwOnFail("zeDeviceGetProperties", zeDeviceGetProperties(_init_structs->getDevice(), &_ze_properties));
+
+    const auto byte_size = ov::element::get_memory_size(_element_type, shape_size(_shape));
+
+    ze_device_external_memory_properties_t desc = {};
+    desc.stype = ZE_STRUCTURE_TYPE_DEVICE_EXTERNAL_MEMORY_PROPERTIES;
+    auto res = zeDeviceGetExternalMemoryProperties(_init_structs->getDevice(), &desc);
+    if (res != ZE_RESULT_SUCCESS ||
+        (desc.memoryAllocationImportTypes != ZE_EXTERNAL_MEMORY_TYPE_FLAG_DMA_BUF &&
+         desc.memoryAllocationImportTypes != ZE_EXTERNAL_MEMORY_TYPE_FLAG_OPAQUE_WIN32_KMT)) {
+        OPENVINO_THROW("Remote tensor functionality is not supported with this driver version");
+    }
+
+    allocate(byte_size);
+}
+
+ZeroRemoteTensor::~ZeroRemoteTensor() {
+    auto res = deallocate();
```

Comment:
Can `deallocate` be called before destructor? If so, it can lead to double free, so a check will be required.

Reply:
I added a check in the deallocate function to check if data is created or not.